iModelCore/Bentley/BeRapidJson/PublicAPI/BeRapidJson/BeJsValue.h

```diff
@@ -834,6 +837,28 @@ struct BeRapidJsonValue : JsValueRef {
     }
     virtual int32_t GetInt(int32_t defVal) const override { return isNumeric() ? m_value->GetInt() : defVal; }
     virtual uint32_t GetUInt(uint32_t defVal) const override { return isNumeric() ? m_value->GetUint() : defVal; }
+    virtual int64_t GetInt64(int64_t defVal) const override { return isNumeric() ? m_value->GetInt64() : defVal; }
+    virtual uint64_t GetUInt64(uint64_t defVal) const override { 
+        if (m_value->IsDouble() && IsLosslessUint64(m_value->GetDouble())) {
+            return static_cast<uint64_t>(m_value->GetDouble());
+        }
+        if (m_value->IsFloat() && IsLosslessUint64(m_value->GetFloat())) {
+            return static_cast<uint64_t>(m_value->GetFloat());
+        }
+        return m_value->IsUint64() ? m_value->GetUint64() : defVal; 
+    }
+    bool IsLosslessUint64(double d) const {
+        if (d < 0.0 || d > static_cast<double>(std::numeric_limits<uint64_t>::max())) {

```

Comment:
I don't understand what definition of "lossless" you are using here. A double cannot losslessly hold every uint64_t value. Don't you want to compare against JS_MAX_SAFE_INTEGER?

Reply:
If it is a double that fits into the range of uint64_t then it can be converted to a uint64 safely was the thought process I had. I added this because when I casted something like 2 to a double, it would not return an id of 2 which I think it should. (Note I don't know 2 being a double is ever realistic)

Reply:
The bug you are trying to accommodate is somebody did the equivalent of:
```
ElementId id;
Json::Value v; // or RapidJson value
v = Json::Value(id.GetValue()); // -> uint64_t
```

That is always stored internally as a uint64_t, losslessly, not as a double. So these `IsLossless` functions will never be involved in your new code path for doubles.

If for some reason somebody's trying to convert a value stored as a **double** into a uint64_t, then that conversion is only accurate if the double value is <= Number.MAX_SAFE_INTEGER, aka the maximal extent of the range of contiguous integer values exactly representable with a 53-bit significand. That isn't a JavaScript thing, it's a IEEE 754 thing.

Our API intentionally makes it difficult to make this mistake (e.g., by making `Json::Value(uint64_t)` constructor explicit). All we want to do in this PR is prevent data corruption due to people having made that mistake.

Reply:
I was assuming that a double of size 2^55 for example could be converted to a uint without any loss, but that is not true?

I think I understand it more the other way around where all uints in the range 2^55-2^56 could not be converted to a double without loss because a double could not represent every single possible number between the range of 2^55-2^56.

Its true in both directions? I assumed that all doubles in the range 2^54->2^63 was a subset of the full range of numbers in the range 2^54->2^63. I also assumed that uint64 can represent the full range of numbers in the range 2^54->2^63).

Reply:
Ok, I see where you are coming from now.